### PR TITLE
Lora unit tests

### PR DESCRIPTION
**Pull request by martinsbruveris**
In addition to adding unit tests I have made the following changes.
- In `LoRADense`, replaced `*args` in `__init__` with an explicit list of parameters. The public TF API is fairly stable and for code inspection it is easier if the PyCharm tooltip shows you the list of parameters.
- In `LoRADense`, renamed `merge_lora_weights()` and `unmerge_lora_weights()` to `merge_weights()` and `unmerge_weights()`. I don't think it is necessary to repeat `lora` here. If possible and within reason, simpler names are preferable.
- In `LoRADense.call()`, moved the `if self.merged:` check to the very beginning of the function, before casting happens. In the merged case we hand everything to the base class.
- Fixed the `LORA_WEIGHT_NAMES` list in `factory.py`.
- Added a `convert_to_lora_layer` helper function to make creating LoRA layers easier.

### CODE REVIEW

File: tests/models/test_lora.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+    # Set non-trivial weights, since bias and kernel_b are zero-initialised.
+    layer.bias = tf.random.uniform(layer.bias.shape)
+    layer.kernel_lora_a = tf.random.uniform(layer.kernel_lora_a.shape)
+    layer.kernel_lora_b = tf.random.uniform(layer.kernel_lora_b.shape)
+
+    x = tf.random.uniform(input_shape)
+    res_1 = layer(x)
+
+    layer.merge_weights()
```
Comment (kevin-keraudren) [inline tests/models/test_lora.py:27]: can you add a check before the call to `merge_weights` that `assert not layer.merged`?

File: tfimm/architectures/lora/factory.py [hunk 2]
```diff
@@ -149,41 +150,97 @@ def convert_to_regular_model(model: tf.keras.Model) -> tf.keras.Model:
...
     if train_bias not in {"none", "all", "lora_only"}:
         raise ValueError(f"Unknown value for train_bias: {train_bias}.")
 
-    # We first set everything to non-trainable
-    model.trainable = False
-
-    # Then we mark LoRA and (optionally) bias layers as trainable
-    for layer in model._flatten_layers(recursive=True, include_self=False):
+    trainable_weights = []
+    for layer in model._flatten_layers(recursive=True, include_self=True):
```
Comment (kevin-keraudren) [inline tfimm/architectures/lora/factory.py:178]: why do we need to include `self`?

File: tfimm/architectures/lora/factory.py [hunk 3]
```diff
@@ -149,41 +150,97 @@ def convert_to_regular_model(model: tf.keras.Model) -> tf.keras.Model:
...
+            tf.keras.layers.DepthwiseConv2D,
+        }
+        and layer.use_bias
+    ):
+        return [layer.bias]
+    elif (
+        type(layer)
+        in {
+            tf.keras.layers.BatchNormalization,
+            tf.keras.layers.LayerNormalization,
```
Comment (kevin-keraudren) [inline tfimm/architectures/lora/factory.py:240]: does it make sense to freeze normalisation layers? if frozen, they will not really "normalise" their inputs?

File: tfimm/architectures/lora/convnext.py [hunk 4]
```diff
@@ -13,37 +14,32 @@
...
 @keras_serializable
 @register_lora_architecture
 class LoRAConvNeXt(ConvNeXt):
     cfg_class = LoRAConvNeXtConfig
-    keys_to_ignore_on_load_missing = ["kernel_lora"]
 
     def __init__(self, cfg: LoRAConvNeXtConfig, **kwargs):
         # We first create the original model
         super().__init__(cfg, **kwargs)
+        self.cfg = cfg
```
[the remark this answers is not in the document]
Comment (martinsbruveris) [reply tfimm/architectures/lora/convnext.py:29]: Yes, it is. But in order for PyCharm to know that `self.cfg` is of type `LoRAConvNeXtConfig` rather than `ConNeXtConfig` I added this line here. Otherwise I get red wavy lines when I access `self.cfg.lora_train_bias`. If there is another way, I'd be happy to change it.

File: tfimm/architectures/lora/factory.py (hunk 2 above)
Comment (martinsbruveris) [reply tfimm/architectures/lora/factory.py:178]: Practically speaking, we don't. But, conceptually, my thinking was: this loop needs to iterate over all objects that can contain variables and that includes the model class itself (since we can call `add_weight` in its build function). So logically it seemed cleaner to include it although I agree that practically I can't imagine a case where it would make a difference.

Technically, we could relax the type hint from `Model` to `Layer` and the function would continue working. And then we could imagine someone calling it on a `LoRADense` layer itself... Something along those lines... But it is all theoretical.

File: tests/models/test_lora.py (hunk 1 above)
Comment (martinsbruveris) [reply tests/models/test_lora.py:27]: Sure.

File: tfimm/architectures/lora/factory.py (hunk 3 above)
Comment (martinsbruveris) [reply tfimm/architectures/lora/factory.py:240]: Normalisation layers contain two parts: First, they normalise their inputs. But then, they contain learnable scale and bias parameters to map the normalised inputs to the desired mean and bias. So here we are saying that during the training we continue to train the bias parameter. Normalization happens regardless.

File: tfimm/architectures/lora/layers.py [hunk 5]
```diff
@@ -73,19 +105,43 @@ def get_config(self):
...
+    Convenience function to convert supported layer types to their LoRA counterparts.
+
+    Args:
+        layer: Layer to be converted.
+        **kwargs: LoRA specific parameters such as ``lora_rank`` have to be passed as
+            kwargs.
+
+    Returns:
+        LoRA layer instance.
+    """
```
[the remark this answers is not in the document]
Comment (martinsbruveris) [reply tfimm/architectures/lora/layers.py:140]: Yup, we can do that.
